### PR TITLE
pkg: add new ring cache to store bucket hot degree

### PR DESCRIPTION
**Pull request by bufferflies**
### What problem does this PR solve?

Issue Number: relate #4726
details: https://leetcode.cn/problems/interval-list-intersections/

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
 None.
```

### CODE REVIEW

File: pkg/cache/ring.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	}
+}
+
+// RingItem is a ring item.
+type RingItem interface {
+	Less(than btree.Item) bool
+	EndKey() []byte
+	// Debris returns the debris after replacing the key range.
+	Debris(startKey, endKey []byte) []RingItem
+	StartKey() []byte
```
Comment (nolouch) [inline pkg/cache/ring.go:41]: move `StartKey() []byte` before`EndKey() []byte`

File: pkg/cache/ring.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+package cache
+
+import (
+	"bytes"
+
+	"github.com/tikv/pd/pkg/btree"
+)
+
+// Ring is a buffer, the key range must be continuous.
```
Comment (nolouch) [inline pkg/cache/ring.go:23]: Why call ring? I misunderstand it as a ring buffer.

File: pkg/cache/ring.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	String() string
+}
+
+// Len returns the length of the ring.
+func (r *Ring) Len() int {
+	return r.tree.Len()
+}
+
+// GetRange returns the items that belong the key range.
+func (r *Ring) GetRange(item RingItem) []RingItem {
```
Comment (nolouch) [inline pkg/cache/ring.go:51]: Add more comments; better to add some ASCII diagrams.

File: pkg/cache/ring.go (hunk 2 above)
Comment (bufferflies) [reply pkg/cache/ring.go:23]: In my opinion, the ring describe all the items are continuous, not allow gaps.

Comment (nolouch) [reply pkg/cache/ring.go]: I think it's not a suitable name, `ring buffer` is a fixed-size buffer.

Comment (bufferflies) [reply pkg/cache/ring.go]: yes, I will rename it as bucket-tree.

File: server/statistics/buckets/bucket_tree.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	Less(than btree.Item) bool
+	StartKey() []byte
+	EndKey() []byte
+	// Debris returns the debris after replacing the key range.
+	Debris(startKey, endKey []byte) []BucketItem
+	String() string
+}
+
+// Len returns the length of the ring.
+func (r *Ring) Len() int {
```
Comment (lhy1024) [inline server/statistics/buckets/bucket_tree.go:46]: Do we also need to rename it?

Comment (bufferflies) [reply server/statistics/buckets/bucket_tree.go]: all fix

File: server/statistics/buckets/bucket_tree.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	var first BucketItem
+	r.tree.DescendLessOrEqual(item, func(i btree.Item) bool {
+		first = i.(BucketItem)
+		return false
+	})
+
+	// find the first item that contains the start key. if not found,
+	// it will use the param.
+	if first == nil || !(bytes.Compare(first.StartKey(), item.StartKey()) <= 0 &&
+		bytes.Compare(item.StartKey(), first.EndKey()) < 0) {
```
Comment (nolouch) [inline server/statistics/buckets/bucket_tree.go:66]: Add a test case about the empty end key; I think you do not handle the cornor case.

**Comment by nolouch** [comment]
Better to directly use `region tree`, we can make an item interface, then you can NewBucketTree and NewRegionTree,  they share a common code.

File: server/core/region_tree.go [hunk 6]
```diff
@@ -26,12 +26,151 @@ import (
 	"go.uber.org/zap"
 )
 
+// RegionItem is region tree item.
+type RegionItem interface {
```
Comment (nolouch) [inline server/core/region_tree.go:30]: RegionItem -> RangeItem
RegionTree -> RTree or `RangeTree`  or else?

better move this part to `pkg`?

Comment (bufferflies) [reply server/core/region_tree.go]: fix

File: pkg/rangetree/range_tree.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+			return true
+		}
+		prev = i.(RangeItem)
+		return false
+	})
+	return prev, next
+}
+
+// GetAt returns the given index item.
+func (r *RangeTree) GetAt(index int) RangeItem {
```
Comment (nolouch) [inline pkg/rangetree/range_tree.go:133]: Where use it?

**Review by nolouch — COMMENTED** [review]
rest LGTM

File: pkg/rangetree/range_tree.go (hunk 7 above)
Comment (bufferflies) [reply pkg/rangetree/range_tree.go:133]: `RandomFollower` and `RandomLeader` will use it.

File: pkg/rangetree/range_tree.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
+package rangetree
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply pkg/rangetree/range_tree.go:15]: fix

File: server/core/region_tree.go [hunk 9]
```diff
@@ -23,19 +23,35 @@ import (
...
+	return r.region.GetStartKey()
+}
+
+// GetEndKey returns the end key of the region.
+func (r *regionItem) GetEndKey() []byte {
+	return r.region.GetEndKey()
+}
+
+// Debris returns nil, don't need to save debris.
+func (r *regionItem) Debris(_, _ []byte) []rangetree.RangeItem {
```
Comment (HunDunDM) [inline server/core/region_tree.go:47]: remove it

**Review by HunDunDM — APPROVED** [review]
rest LGTM

File: server/core/region_tree.go (hunk 9 above)
Comment (bufferflies) [reply server/core/region_tree.go]: fix

**Review by nolouch — APPROVED** [review]
